examples/llm/src/models/layers/moe/tutel_moe.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+import warnings
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+from composer.utils import dist
+from omegaconf import DictConfig
+from omegaconf.listconfig import ListConfig
+from tutel import moe as tutel_moe
+from tutel.experts.ffn import FusedExpertsNetwork
+from tutel.gates.cosine_top import CosineTopKGate
+from tutel.gates.top import LinearTopKGate
+from tutel.impls.moe_layer import MOELayer
+
+from examples.common.seed_ctx_manager import SeedContextManager
+from examples.llm.src.models.layers.moe.base_moe import BaseMoE
+
+
+class TutelMOE(BaseMoE):
+    """Wrap and configure tutel.moe."""
+
+    def __init__(self,
+                 cfg: DictConfig,
+                 block_idx: int,
+                 device: Optional[str] = None):
+        super().__init__()
+
+        world_size = dist.get_world_size()
+
+        num_experts = cfg.moe.get('num_experts')
+        if isinstance(num_experts, ListConfig):
+            # enables pyramid moe
+            num_experts = num_experts[block_idx]
```

Comment:
PyramidMoE is implemented, and the one test I had runs a PyramidMoE.
I'll add a variant that runs a regular moe.

Reply:
Would the right way to test for behavior not impl for PyramidMoEs, be to somehow verify that the MoE lib being used doesn't require that at every depth the user inits the same number of experts?
This sounds heavily based on MoE lib implementation.

Reply:
I was just thinking it would be nice to just try using num_experts has `__getitem__` rather than checking that it's a particular class because "[depend on abstractions, not concretions](https://softwareengineering.stackexchange.com/questions/401769/depend-on-abstractions-not-on-concretions-what-is-the-exact-meaning-of-this-t)". Something like:
```suggestion
        if isinstance(num_experts, Sequence):
            # enables pyramid moe
            num_experts = num_experts[block_idx]
```
which counts since `isinstance(..., Sequence)` tests for an interface, not a concrete class. Or, using the principle that in Python it's [better to ask for forgiveness than permission](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/), we could:
```python
try:
    num_experts = num_experts[block_idx]
except TypeError:
    pass
```
although personally I think this is clunkier.